transport/src/main/java/io/netty/channel/AbstractChannel.java

```diff
@@ -650,38 +650,19 @@ private void shutdownOutput(final ChannelPromise promise, Throwable cause) {
             final Throwable shutdownCause = cause == null ?
                     new ChannelOutputShutdownException("Channel output shutdown") :
                     new ChannelOutputShutdownException("Channel output shutdown", cause);
-            Executor closeExecutor = prepareToClose();
-            if (closeExecutor != null) {
-                closeExecutor.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        try {
-                            // Execute the shutdown.
-                            doShutdownOutput();
-                            promise.setSuccess();
-                        } catch (Throwable err) {
-                            promise.setFailure(err);
-                        } finally {
-                            // Dispatch to the EventLoop
-                            eventLoop().execute(new Runnable() {
-                                @Override
-                                public void run() {
-                                    closeOutboundBufferForShutdown(pipeline, outboundBuffer, shutdownCause);
-                                }
-                            });
-                        }
-                    }
-                });
-            } else {
-                try {
-                    // Execute the shutdown.
-                    doShutdownOutput();
-                    promise.setSuccess();
-                } catch (Throwable err) {
-                    promise.setFailure(err);
-                } finally {
-                    closeOutboundBufferForShutdown(pipeline, outboundBuffer, shutdownCause);
-                }
+
+            // When a side enables SO_LINGER and calls showdownOutput(...) to start TCP half-closure, we can not call doDeregister here
+            // because we should ensure this side in fin_wait2 state can still receive and process the data which is send by another side in the close_wait state。
+            // See https://github.com/netty/netty/issues/11981
+            try {
```

Comment:
@huibinliupush I am sure you will need to have the same change in the other transports as well (Epoll, KQueue).

Reply:
OK, i'll check it  now

Reply:
> @huibinliupush I am sure you will need to have the same change in the other transports as well (Epoll, KQueue).

@normanmaurer OK, i'll check it right now

Reply:
@normanmaurer 

i have checked `EpollSocketChannel` and `KQueueSocketChannel` again，they all call 
`((AbstractUnsafe) unsafe()).shutdownOutput(promise)` method to shutdownOutput In their parent class

Am I missing something?

Reply:
ah ok... Sorry I thought they also offload :)